Cleanups in typed TestProbe fishForMessage impl

dwijnand opened the pull request.

dwijnand committed:
Guarantee match exhaustion in fishForMessage

... by introducing FishingOutcome.ContinueOutcome to the type hierarchy.

dwijnand committed:
Simplify some timeout calculation in fishForMessage

'timeout' is always finite, as it's a `FiniteDuration`, and, even if it
weren't, the calculation is safe for non-finite durations.

dwijnand committed:
Avoid double handing of timeouts in fishForMessage

If 'newTimeout' is sub-zero, just loop again which will trigger the
other throw.

dwijnand committed:
Tighten MatchError handling in fishForMessage

Avoid a whole bunch of other code being accidentally handled.

Also it allows for the inner 'loop' method to be @tailrec (next commit).

dwijnand committed:
Make fishForMessage's inner loop @tailrec!

Apparently @tailrec and try/catch (with a re-throw in the catch) don't
mix well.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -57,7 +57,7 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
   private var lastWasNoMessage = false
 
-  private var lastMessage: Option[M] = None
+  private var lastMessage: OptionVal[M] = OptionVal.none
```
`OptionVal` is intended for places where extreme performance is needed, when we don't want to allocate a new instance for `Some`. That can't be a requirement for the testkit so we should use standard `Option`, unless there is another reason I have missed here?

dwijnand replied:
I'm happy with `Option` if you are.

dwijnand replied:
Oh, wait, here I'm switching from `Option` to `OptionVal` (not replacing `null` usage, that's another change).  If it's not needed I'm happy to drop it.

patriknw commented on akka-actor/src/main/scala/akka/util/OptionVal.scala:
```diff
@@ -9,13 +9,15 @@ package akka.util
   }
 
+  def none[A]: OptionVal[A] = None.asInstanceOf[OptionVal[A]]
+
```
That looks too simple, but fantastic if it works.
The purpose is to be allocation free for both creating `Some(a)` and for
```
opt match {
  case OptionVal.Some(a) =>
  case OptionVal.None =>
}
```

dwijnand replied:
Yep, it should all work as before.

One thing that's different (I mean, even before my change) is there's no match exhaustion, so if nothing saves you from forgetting the `OptionVal.Some` or `OptionVal.None` case.  And sometimes you can have duplicates, like multiple `case OptionVal.None` 😄

dwijnand replied:
Moved to #25868

dwijnand committed:
Make receiveOne return Option[M] rather than nullable M

Avoid future users from forgetting to consider the null case, which
happened in fishForMessage_internal.

dwijnand commented:
Dropped switch to `OptionVal`, which removed the need for the change to `OptionVal`.  So I've proposed that separately in #25868.

Also rebased to resolve the (self-inflicted!) merge conflict.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -133,9 +133,9 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-    assert(obj == o, s"expected $obj, found $o" + hintOrEmptyString)
-    o.asInstanceOf[T]
+    assert(o.isDefined, s"timeout ($max) during expectMessage while waiting for $obj" + hintOrEmptyString)
+    assert(obj == o.get, s"expected $obj, found $o" + hintOrEmptyString)
```
`found ${o.get}`
smells like these two asserts and the final return should be rewritten as a `match`.
Note that you can `throw new AssertionError` instead of using `assert`.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -169,49 +170,38 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-    assert(BoxedType(c) isInstance o, s"expected $c, found ${o.getClass} ($o)")
-    o.asInstanceOf[C]
+    assert(o.isDefined, s"timeout ($max) during expectMessageClass waiting for $c")
+    assert(BoxedType(c) isInstance o.get, s"expected $c, found ${o.get.getClass} (${o.get})")
```
match here also

patriknw reviewed (commented):
looking good

dwijnand committed:
Switch to pattern matching Options

dwijnand committed:
Cast properly

Thanks, type inferencer, for inferring Nothing there! o_O

chbatey commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -169,49 +173,44 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
   override protected def fishForMessage_internal(max: FiniteDuration, hint: String, fisher: M ⇒ FishingOutcome): List[M] = {
-    // not tailrec but that should be ok
-    def loop(timeout: FiniteDuration, seen: List[M]): List[M] = {
+    @tailrec def loop(timeout: FiniteDuration, seen: List[M]): List[M] = {
```
:+1:

chbatey reviewed (approved):
LGTM, I don't mind regarding the MatchError. Fish isn't meant to always take a PartialFunction, it is just a convenience if you want to pass one in.

dwijnand committed:
Make fish's PartialFunction convenience explicit

dwijnand commented:
Ah, I see.  I can see the convenience of having this, actually.  (And I'll use it too!)  I pushed a commit to make the convenience intent more explicit.

dwijnand committed:
Update the scaladsl fishForMessage Scaladocs

dwijnand commented:
Amended the Scaladoc to reflect that we only handle `PartialFunction` now, rather than any function throwing `MatchError`.  Also dropped the incorrect part about type `T`, which doesn't exist.

dwijnand committed:
Restore MatchError catching logic & docs

TIL partial functions aren't of class PartialFunction anymore, they're
lambda classes.

dwijnand commented:
Learnt some scalac internal details.  High five to whoever wrote that test.

dwijnand commented:
Good to go.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -169,49 +173,44 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
           }
+          outcome match {
+            case FishingOutcome.Complete    ⇒ (message :: seen).reverse
+            case FishingOutcome.Fail(error) ⇒ throw new AssertionError(s"$error, hint: $hint")
```
is this intended to not use `assertFail`?

dwijnand replied:
Yes because of the leading "assertion failed: " that gets added with the old `assert` conditions.  But that's not so important, so I'll reuse `assertFail` more.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -133,9 +133,12 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-    assert(obj == o, s"expected $obj, found $o" + hintOrEmptyString)
-    o.asInstanceOf[T]
+    o match {
+      case Some(o) if obj == o ⇒ o.asInstanceOf[T]
```
somewhat confusing to hide the outer `o`

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -133,9 +133,12 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-    o.asInstanceOf[T]
+    o match {
+      case Some(o) if obj == o ⇒ o.asInstanceOf[T]
+      case Some(o)             ⇒ assertFail(s"expected $obj, found $o$hintOrEmptyString")
```
same here ^

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -169,49 +173,44 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-    lastWasNoMessage = true
+    o match {
+      case None    ⇒ lastWasNoMessage = true
+      case Some(o) ⇒ assertFail(s"received unexpected message $o")
```
and here ^ and a few more places

patriknw reviewed (commented):
looking good

dwijnand committed:
Tweak assertFail so it can be used more

dwijnand committed:
Avoid name shadowing

... on request. (I <3 name shadowing)

dwijnand committed:
Fix formatting

dwijnand commented:
I can resubmit this if it makes it easier to review and land.

patriknw reviewed (approved):
LGTM, thanks
